File: JavaScript/Co-op/functional.js
//- Function to simulate 'union' operation
function union(leftSet, rightSet, dims) {
    const mergedSet = [...leftSet, ...rightSet];
    return [...new Set(mergedSet.map(JSON.stringify))].map(JSON.parse);
  }
  
//- Function to simulate 'difference' operation
function difference(leftSet, rightSet, dims) {
    const rightSetStr = new Set(rightSet.map(JSON.stringify));
    return leftSet.filter(item => rightSetStr.has(JSON.stringify(item)));
  }
  
//- Function to simulate 'intersection' operation
function intersection(leftSet, rightSet, dims) {
    const allItems = [...leftSet, ...rightSet, ...rightSet];
    const itemCounts = {};
    
    allItems.forEach(item => {
      const itemStr = JSON.stringify(item);
      itemCounts[itemStr] = (itemCounts[itemStr] || 0) + 1;
    });
  
    return Object.keys(itemCounts).filter(itemStr => itemCounts[itemStr] === 1).map(JSON.parse);
  }
  
  
  
//- Function to simulate 'complement' operation
function complement(data, inputSet, dims) {
    const allItems = [...data, ...inputSet];
    const itemCounts = {};
    
    allItems.forEach(item => {
      const itemStr = JSON.stringify(item);
      itemCounts[itemStr] = (itemCounts[itemStr] || 0) + 1;
    });
  
    return Object.keys(itemCounts).filter(itemStr => itemCounts[itemStr] === 1).map(JSON.parse);
  }
  
//* OPTIONAL FUNCTION TO PRINT DATAFRAME
function printDataFrame(data, label) {
    console.log(label);
    const keys = Object.keys(data[0]);
    console.log(keys.join(" | "));
    console.log("-".repeat(keys.join(" | ").length));
    for (const row of data) {
        console.log(Object.values(row).join(" | "));
    }
    console.log('\n');
  }

File: Python/Co-op/operators/functional.py
import pandas as pd

#- Function to simulate 'union' operation
def union(left_set, right_set, dims):
    return pd.merge(left_set, right_set, on=dims, how='outer')

#- Function to simulate 'difference' operation
def difference(left_set, right_set, dims):
    return pd.merge(left_set, right_set, on=dims)

#- Function to simulate 'intersection' operation
def intersection(left_set, right_set, dims):
    left_set = left_set.append(right_set)
    left_set = left_set.append(right_set)
    return left_set.drop_duplicates(subset=dims, keep=False)

#- Function to simulate 'complement' operation
def complement(data, input_set, dims):
    data = data.append(input_set)
    return data.drop_duplicates(subset=dims, keep=False)
